File: src/trading/strategies/stop-loss-arb/algo.ts
import {writeJSONFile, jsonPrettyPrint, renameFile} from '../../../utils/file';
import {FloatCalculator as fc} from '../../../utils/float-calculator';
import {log} from '../../../utils/log';
import {getSimulatedSnapshot, isLiveTrading} from '../../../utils/price-simulator';
import {
    getCurrentMomentInNewYork,
    getCurrentTimeStamp,
    getMomentForTime,
    kTradingEndTime,
} from '../../../utils/time';
import {
    Snapshot,
    OrderAction,
    BrokerageClient,
} from '../../brokerage-clients/brokerage-client';
import {getStockStateFilePath} from './state';
import {IntervalType, StockState} from './types';

export async function reconcileStockPosition(
    stock: string,
    stockState: StockState,
    date: string,
    brokerageClient?: BrokerageClient,
): Promise<{
    snapshot: Snapshot;
    crossedThreshold: boolean;
}> {
    // 0)
    let snapshot: Snapshot;
    snapshot = await (brokerageClient ?
        brokerageClient.getSnapshot(stockState.brokerageId) :
        getSimulatedSnapshot(stockState));

    const isSnapshotChanged = isSnapshotChange(snapshot, stockState);
    const isPastTradingTimeBool = isPastTradingTime();
    if (isSnapshotChanged || isPastTradingTimeBool) {
        updateSnaphotOnState(stockState, snapshot);
        updateExitPnL(stockState);

        const isExitPnlBeyondThresholdsStr = isExitPnlBeyondThresholds(stockState);
        if (
            isLiveTrading() &&
            (isExitPnlBeyondThresholdsStr || isPastTradingTimeBool)
        ) {
            if (stockState.position !== 0) {
                await setNewPosition({
                    stock,
                    brokerageClient,
                    stockState,
                    newPosition: 0,
                    snapshot,
                });

                setRealizedPnL(stockState);
            }

            await writeJSONFile(
                getStockStateFilePath(stock, date),
                jsonPrettyPrint(stockState),
            );

            const oldPath = getStockStateFilePath(stock, date);

            const suffix = isExitPnlBeyondThresholdsStr || 'N';
            const newFileName = `_${stock}_${suffix}`;
            const newPath = getStockStateFilePath(newFileName, date);

            await renameFile(oldPath, newPath);

            return {
                snapshot,
                crossedThreshold: true,
            };
        }
    }

    if (isWideBidAskSpread(snapshot, stockState) || !snapshot.bid || !snapshot.ask) {
        return {
            snapshot,
            crossedThreshold: false,
        };
    }

    // 1)
    checkCrossings(stockState, snapshot);

    // 2)
    let intervalIndicesToExecute = getNumToBuy(stockState, snapshot);
    const numToBuy = intervalIndicesToExecute.length;

    // 3)
    let numToSell = 0;
    if (numToBuy === 0) {
        intervalIndicesToExecute = getNumToSell(stockState, snapshot);
        numToSell = intervalIndicesToExecute.length;
    }

    // 4)
    let newPosition: number | undefined;
    if (numToBuy > 0) {
        newPosition = stockState.position + stockState.sharesPerInterval * numToBuy;
    } else if (numToSell > 0) {
        newPosition = stockState.position - stockState.sharesPerInterval * numToSell;
    }

    // 5)
    if (newPosition !== undefined) {
        await setNewPosition({
            stock,
            brokerageClient,
            stockState,
            newPosition,
            snapshot,
        });

        checkCrossings(stockState, snapshot);
    }

    // 6)
    if (isSnapshotChanged) {
        if (isLiveTrading()) {
            await writeJSONFile(
                getStockStateFilePath(stock, date),
                jsonPrettyPrint(stockState),
            );
        }
    }

    return {
        snapshot,
        crossedThreshold: false,
    };
}

function isExitPnlBeyondThresholds(stockState: StockState): string {
    if (isLiveTrading() && stockState.profitThreshold && stockState.lossThreshold) {
        if (fc.gte(stockState.exitPnLAsPercentage, stockState.profitThreshold)) {
            return 'W';
        }

        if (fc.lte(stockState.exitPnLAsPercentage, stockState.lossThreshold)) {
            return 'L';
        }
    }

    return '';
}

function isPastTradingTime(): boolean {
    // return false;

    const tradingEndTime = getMomentForTime(kTradingEndTime);
    const currentMomentInNewYork = getCurrentMomentInNewYork();

    const isPastTradingTimeBool = currentMomentInNewYork.isSameOrAfter(tradingEndTime);

    return isPastTradingTimeBool;
}

function isWideBidAskSpread({bid, ask}: Snapshot, stockState: StockState): boolean {
    // return false;
    return fc.gte(fc.subtract(ask, bid), stockState.spaceBetweenIntervals) === 1;
}

function checkCrossings(stockState: StockState, {bid, ask}: Snapshot): void {
    const {intervals} = stockState;

    for (const interval of intervals) {
        if (
            interval[OrderAction.BUY].active &&
            !interval[OrderAction.BUY].crossed &&
            fc.lt(ask, interval[OrderAction.BUY].price)
        ) {
            interval[OrderAction.BUY].crossed = true;
        }

        if (
            interval[OrderAction.SELL].active &&
            !interval[OrderAction.SELL].crossed &&
            fc.gt(bid, interval[OrderAction.SELL].price)
        ) {
            interval[OrderAction.SELL].crossed = true;
        }
    }
}

function getNumToBuy(stockState: StockState, {ask}: Snapshot): number[] {
    const {intervals, position} = stockState;

    let newPosition = position;
    const indicesToExecute: number[] = [];
    for (let i = intervals.length - 1; i >= 0; i--) {
        const interval = intervals[i];

        if (
            fc.gte(ask, interval[OrderAction.BUY].price) &&
            interval[OrderAction.BUY].active &&
            interval[OrderAction.BUY].crossed
        ) {
            if (newPosition < interval.positionLimit) {
                indicesToExecute.unshift(i);
                newPosition += stockState.sharesPerInterval;
            }
        }
    }

    if (stockState.isStaticIntervals) {
        addSkippedBuysIfRequired(stockState, indicesToExecute);
    }

    for (const index of indicesToExecute) {
        const interval = intervals[index];

        interval[OrderAction.BUY].active = false;
        interval[OrderAction.BUY].crossed = false;

        interval[OrderAction.SELL].active = true;
        interval[OrderAction.SELL].crossed = false;
    }

    if (indicesToExecute.length > 0) {
        if (!stockState.isStaticIntervals) {
            correctBadBuyIfRequired(stockState, indicesToExecute);
        }
    }

    return indicesToExecute;
}

function getNumToSell(stockState: StockState, {bid}: Snapshot): number[] {
    const {intervals, position} = stockState;

    let newPosition = position;
    const indicesToExecute: number[] = [];
    for (const [i, interval] of intervals.entries()) {
        if (
            fc.lte(bid, interval[OrderAction.SELL].price) &&
            interval[OrderAction.SELL].active &&
            interval[OrderAction.SELL].crossed
        ) {
            if (newPosition > interval.positionLimit) {
                indicesToExecute.push(i);
                newPosition -= stockState.sharesPerInterval;
            }
        }
    }

    if (stockState.isStaticIntervals) {
        addSkippedSellsIfRequired(stockState, indicesToExecute);
    }

    for (const index of indicesToExecute) {
        const interval = intervals[index];

        interval[OrderAction.SELL].active = false;
        interval[OrderAction.SELL].crossed = false;

        interval[OrderAction.BUY].active = true;
        interval[OrderAction.BUY].crossed = false;
    }

    if (indicesToExecute.length > 0) {
        if (!stockState.isStaticIntervals) {
            correctBadSellIfRequired(stockState, indicesToExecute);
        }
    }

    return indicesToExecute;
}

function isSnapshotChange(snapshot: Snapshot, stockState: StockState): boolean {
    if (!stockState.lastAsk || !stockState.lastBid) {
        return true;
    }

    return (
        !fc.eq(stockState.lastAsk, snapshot.ask) ||
        !fc.eq(stockState.lastBid, snapshot.bid)
    );
}

async function setNewPosition({
    stock,
    brokerageClient,
    stockState,
    newPosition,
    snapshot,
}: {
    stock: string;
    brokerageClient?: BrokerageClient;
    stockState: StockState;
    newPosition: number;
    snapshot: Snapshot;
}): Promise<void> {
    const previousPosition = stockState.position;
    stockState.position = newPosition;

    const orderSide: OrderAction =
        newPosition > previousPosition ? OrderAction.BUY : OrderAction.SELL;

    const quotedPrice = orderSide === OrderAction.BUY ? snapshot.ask : snapshot.bid;
    let priceSetAt: number = quotedPrice;

    const tradingLog: (typeof stockState.tradingLogs)[number] = {
        action: orderSide,
        timeStamp: snapshot.timestamp || getCurrentTimeStamp(),
        quotedPrice,
        realizedPrice: priceSetAt,
        previousPosition,
        newPosition,
    };
    stockState.tradingLogs.push(tradingLog);

    if (brokerageClient) {
        priceSetAt = await brokerageClient.setSecurityPosition({
            brokerageIdOfSecurity: stockState.brokerageId,
            currentPosition: previousPosition * stockState.numContracts,
            newPosition: newPosition * stockState.numContracts,
        });

        log(
            `Changed position for ${stock}, action: ${orderSide}, fullNewPosition: ${
                newPosition * stockState.numContracts
            }, quotedPrice: ${tradingLog.quotedPrice}, realizedPrice: ${
                tradingLog.realizedPrice
            }`,
        );

        await writeJSONFile(
            getStockStateFilePath(stock, stockState.date),
            jsonPrettyPrint(stockState),
        );
    }

    const newNetPositionValue = getNewNetPositionValue({
        currentPositionValue: stockState.netPositionValue,
        commissionPerShare: stockState.brokerageTradingCostPerShare,
        orderSide,
        newPosition,
        previousPosition,
        priceSetAt,
    });

    stockState.netPositionValue = newNetPositionValue;
}

function getNewNetPositionValue({
    currentPositionValue,
    commissionPerShare,
    orderSide,
    newPosition,
    previousPosition,
    priceSetAt,
}: {
    currentPositionValue: number;
    commissionPerShare: number;
    orderSide: OrderAction;
    newPosition: number;
    previousPosition: number;
    priceSetAt: number;
}): number {
    const quantity = Math.abs(newPosition - previousPosition);

    const commissionCosts = fc.multiply(quantity, commissionPerShare);

    let change = -commissionCosts;

    const orderValue = fc.multiply(quantity, priceSetAt);

    if (orderSide === OrderAction.BUY) {
        change = fc.subtract(change, orderValue);
    } else if (orderSide === OrderAction.SELL) {
        change = fc.add(change, orderValue);
    }

    const newPositionValue = fc.add(currentPositionValue, change);

    return newPositionValue;
}

function setRealizedPnL(stockState: StockState): void {
    if (stockState.position !== 0) {
        throw new Error('Cannot set realized PnL because Position is not zero');
    }

    const percentageDenominator = fc.multiply(
        stockState.targetPosition + stockState.sharesPerInterval,
        stockState.initialPrice,
    );

    const realizedPnLAsPercentage = fc.multiply(
        fc.divide(stockState.netPositionValue, percentageDenominator),
        100,
    );

    stockState.realizedPnLAsPercentage = realizedPnLAsPercentage;
}

function updateSnaphotOnState(stockState: StockState, snapshot: Snapshot): void {
    stockState.lastAsk = snapshot.ask;
    stockState.lastBid = snapshot.bid;
}

function updateExitPnL(stockState: StockState): void {
    const {lastAsk, lastBid, position} = stockState;

    if (position === 0) {
        return;
    }

    const orderSide: OrderAction = position > 0 ? OrderAction.SELL : OrderAction.BUY;

    const priceSetAt = orderSide === OrderAction.BUY ? lastAsk : lastBid;

    const ifClosingPositionValue = getNewNetPositionValue({
        currentPositionValue: stockState.netPositionValue,
        commissionPerShare: stockState.brokerageTradingCostPerShare,
        orderSide,
        newPosition: 0,
        previousPosition: position,
        priceSetAt,
    });

    const percentageDenominator = fc.multiply(
        stockState.targetPosition + stockState.sharesPerInterval,
        stockState.initialPrice,
    );

    const exitPnLAsPercentage = fc.multiply(
        fc.divide(ifClosingPositionValue, percentageDenominator),
        100,
    );

    stockState.exitPnLAsPercentage = exitPnLAsPercentage;

    if (fc.gt(exitPnLAsPercentage, stockState.maxMovingProfitAsPercentage)) {
        stockState.maxMovingProfitAsPercentage = exitPnLAsPercentage;
    }

    if (fc.lt(exitPnLAsPercentage, stockState.maxMovingLossAsPercentage)) {
        stockState.maxMovingLossAsPercentage = exitPnLAsPercentage;
    }

    // Here the cpp version is different, because it has the reched_when losses...
}

export function reconcileRealizedPnlWhenHistoricalSnapshotsExhausted(
    stockState: StockState,
): void {
    updateExitPnL(stockState);

    setNewPosition({
        stock: stockState.brokerageId,
        brokerageClient: undefined,
        stockState,
        newPosition: 0,
        snapshot: {
            ask: stockState.lastAsk,
            bid: stockState.lastBid,
            timestamp: getCurrentTimeStamp(),
        },
    });

    setRealizedPnL(stockState);
}

function correctBadBuyIfRequired(
    stockState: StockState,
    indexesToExecute: number[],
): void {
    const lowestIndexExecuted = indexesToExecute[indexesToExecute.length - 1];
    if (lowestIndexExecuted >= stockState.intervals.length - 1) {
        return;
    }

    const intervalBelowLowestIntervalExecuted =
        stockState.intervals[lowestIndexExecuted + 1];
    if (!intervalBelowLowestIntervalExecuted[OrderAction.BUY].active) {
        return;
    }

    intervalBelowLowestIntervalExecuted[OrderAction.BUY].active = false;
    intervalBelowLowestIntervalExecuted[OrderAction.BUY].crossed = false;
    intervalBelowLowestIntervalExecuted[OrderAction.SELL].active = true;
    intervalBelowLowestIntervalExecuted[OrderAction.SELL].crossed = false;

    const topIntervalExecuted = stockState.intervals[indexesToExecute[0]];
    topIntervalExecuted[OrderAction.BUY].active = true;
    topIntervalExecuted[OrderAction.BUY].crossed = false;
    topIntervalExecuted[OrderAction.SELL].active = false;
    topIntervalExecuted[OrderAction.SELL].crossed = false;

    for (const interval of stockState.intervals) {
        interval[OrderAction.BUY].price = fc.add(
            interval[OrderAction.BUY].price,
            stockState.spaceBetweenIntervals,
        );
        interval[OrderAction.SELL].price = fc.add(
            interval[OrderAction.SELL].price,
            stockState.spaceBetweenIntervals,
        );
    }
}

function correctBadSellIfRequired(
    stockState: StockState,
    indexesToExecute: number[],
): void {
    const highestIndexExecuted = indexesToExecute[0];
    if (highestIndexExecuted === 0) {
        return;
    }

    const intervalAboveHighestIntervalExecuted =
        stockState.intervals[highestIndexExecuted - 1];
    if (!intervalAboveHighestIntervalExecuted[OrderAction.SELL].active) {
        return;
    }

    intervalAboveHighestIntervalExecuted[OrderAction.SELL].active = false;
    intervalAboveHighestIntervalExecuted[OrderAction.SELL].crossed = false;
    intervalAboveHighestIntervalExecuted[OrderAction.BUY].active = true;
    intervalAboveHighestIntervalExecuted[OrderAction.BUY].crossed = false;

    const bottomIntervalExecuted =
        stockState.intervals[indexesToExecute[indexesToExecute.length - 1]];
    bottomIntervalExecuted[OrderAction.SELL].active = true;
    bottomIntervalExecuted[OrderAction.SELL].crossed = false;
    bottomIntervalExecuted[OrderAction.BUY].active = false;
    bottomIntervalExecuted[OrderAction.BUY].crossed = false;

    for (const interval of stockState.intervals) {
        interval[OrderAction.BUY].price = fc.subtract(
            interval[OrderAction.BUY].price,
            stockState.spaceBetweenIntervals,
        );
        interval[OrderAction.SELL].price = fc.subtract(
            interval[OrderAction.SELL].price,
            stockState.spaceBetweenIntervals,
        );
    }
}

function addSkippedBuysIfRequired(
    stockState: StockState,
    indexesToExecute: number[],
): void {
    if (indexesToExecute.length === 0) {
        return;
    }

    const {intervals} = stockState;
    const bottomOriginalIndexToExecute = indexesToExecute[indexesToExecute.length - 1];
    for (let i = intervals.length - 1; i > bottomOriginalIndexToExecute; i--) {
        const interval = intervals[i];

        if (interval[OrderAction.BUY].active) {
            indexesToExecute.push(i);
        }
    }
}

function addSkippedSellsIfRequired(
    stockState: StockState,
    indexesToExecute: number[],
): void {
    if (indexesToExecute.length === 0) {
        return;
    }

    const {intervals} = stockState;
    const topOriginalIndexToExecute = indexesToExecute[0];
    for (let i = 0; i < topOriginalIndexToExecute; i++) {
        const interval = intervals[i];

        if (interval[OrderAction.SELL].active) {
            indexesToExecute.unshift(i);
        }
    }
}

File: src/trading/strategies/stop-loss-arb/cpp/algo.cpp

#include "algo.hpp"

#include <vector>

#include "price_simulator.hpp"
#include "types.hpp"

using namespace std;

Snapshot ReconcileStockPosition(const std::string& stock, StockState& stockState)
{
    // 0)
    Snapshot snapshot = GetSimulatedSnapshot(stockState);

    const bool isSnapshotChanged = IsSnapshotChange(snapshot, stockState);
    if (isSnapshotChanged)
    {
        UpdateSnaphotOnState(stockState, snapshot);
        UpdateExitPnL(stockState);
    }

    if (IsWideBidAskSpread(snapshot, stockState) || !snapshot.bid || !snapshot.ask)
    {
        return snapshot;
    }

    // 1)
    CheckCrossings(stockState, snapshot);

    // 2)
    vector<int> intervalIndicesToExecute = GetNumToBuy(stockState, snapshot);
    const int numToBuy = intervalIndicesToExecute.size();

    // 3)
    int numToSell = 0;
    if (numToBuy == 0)
    {
        intervalIndicesToExecute = GetNumToSell(stockState, snapshot);
        numToSell = intervalIndicesToExecute.size();
    }

    // 4)
    optional<int> newPosition;
    if (numToBuy > 0)
    {
        newPosition = stockState.position + stockState.sharesPerInterval * numToBuy;
    }
    else if (numToSell > 0)
    {
        newPosition = stockState.position - stockState.sharesPerInterval * numToSell;
    }

    // 5)
    if (newPosition.has_value())
    {
        SetNewPosition(stock, stockState, newPosition.value(), snapshot);

        CheckCrossings(stockState, snapshot);
    }

    // 6)
    // No Step 6 on cpp because we've decided not to add Live Trading logic

    return snapshot;
}

bool IsWideBidAskSpread(const Snapshot& snapshot, const StockState& stockState)
{
    return (snapshot.ask - snapshot.bid) >= stockState.spaceBetweenIntervals;
}

void CheckCrossings(StockState& stockState, const Snapshot& snapshot)
{
    auto& intervals = stockState.intervals;

    for (auto& interval : intervals)
    {
        if (interval.BUY.active && !interval.BUY.crossed &&
            snapshot.ask < interval.BUY.price)
        {
            interval.BUY.crossed = true;
        }

        if (interval.SELL.active && !interval.SELL.crossed &&
            snapshot.bid > interval.SELL.price)
        {
            interval.SELL.crossed = true;
        }
    }
}

std::vector<int> GetNumToBuy(StockState& stockState, const Snapshot& snapshot)
{
    auto& intervals = stockState.intervals;
    int position = stockState.position;

    int newPosition = position;
    vector<int> indicesToExecute;

    for (int i = intervals.size() - 1; i >= 0; --i)
    {
        const auto& interval = intervals[i];

        if (snapshot.ask >= interval.BUY.price && interval.BUY.active &&
            interval.BUY.crossed)
        {
            if (newPosition < interval.positionLimit)
            {
                indicesToExecute.insert(indicesToExecute.begin(), i);
                newPosition += stockState.sharesPerInterval;
            }
        }
    }

    if (stockState.isStaticIntervals)
    {
        AddSkippedBuysIfRequired(stockState, indicesToExecute);
    }

    for (const int index : indicesToExecute)
    {
        auto& interval = intervals[index];

        interval.BUY.active = false;
        interval.BUY.crossed = false;

        interval.SELL.active = true;
        interval.SELL.crossed = false;
    }

    if (!indicesToExecute.empty())
    {
        if (!stockState.isStaticIntervals)
        {
            CorrectBadBuyIfRequired(stockState, indicesToExecute);
        }
    }

    return indicesToExecute;
}

std::vector<int> GetNumToSell(StockState& stockState, const Snapshot& snapshot)
{
    auto& intervals = stockState.intervals;
    int position = stockState.position;

    int newPosition = position;
    vector<int> indicesToExecute;

    for (int i = 0; i < static_cast<int>(intervals.size()); ++i)
    {
        const auto& interval = intervals[i];

        if (snapshot.bid <= interval.SELL.price && interval.SELL.active &&
            interval.SELL.crossed)
        {
            if (newPosition > interval.positionLimit)
            {
                indicesToExecute.push_back(i);
                newPosition -= stockState.sharesPerInterval;
            }
        }
    }

    if (stockState.isStaticIntervals)
    {
        AddSkippedSellsIfRequired(stockState, indicesToExecute);
    }

    for (const int index : indicesToExecute)
    {
        auto& interval = intervals[index];

        interval.SELL.active = false;
        interval.SELL.crossed = false;

        interval.BUY.active = true;
        interval.BUY.crossed = false;
    }

    if (!indicesToExecute.empty())
    {
        if (!stockState.isStaticIntervals)
        {
            CorrectBadSellIfRequired(stockState, indicesToExecute);
        }
    }

    return indicesToExecute;
}

bool IsSnapshotChange(const Snapshot& snapshot, const StockState& stockState)
{
    if (!stockState.lastAsk || !stockState.lastBid)
    {
        return true;
    }

    return stockState.lastAsk != snapshot.ask || stockState.lastBid != snapshot.bid;
}

void SetNewPosition(
    const std::string& stock,
    StockState& stockState,
    const int& newPosition,
    const Snapshot& snapshot
)
{
    const auto previousPosition = stockState.position;
    stockState.position = newPosition;

    const string orderSide = newPosition > previousPosition ? "BUY" : "SELL";
    const Decimal quotedPrice =
        orderSide == static_cast<string>("BUY") ? snapshot.ask : snapshot.bid;

    const auto newNetPositionValue = GetNewNetPositionValue(
        stockState.netPositionValue,
        stockState.brokerageTradingCostPerShare,
        orderSide,
        newPosition,
        previousPosition,
        quotedPrice
    );

    stockState.netPositionValue = newNetPositionValue;
}

Decimal GetNewNetPositionValue(
    const Decimal& currentPositionValue,
    const Decimal& commissionPerShare,
    const std::string& orderSide,
    const int& newPosition,
    const int& previousPosition,
    const Decimal& priceSetAt
)
{
    int quantity = abs(newPosition - previousPosition);

    Decimal commissionCosts = Decimal(quantity) * commissionPerShare;

    Decimal change = -commissionCosts;

    Decimal orderValue = Decimal(quantity) * priceSetAt;

    if (orderSide == "BUY")
    {
        change -= orderValue;
    }
    else if (orderSide == "SELL")
    {
        change += orderValue;
    }

    Decimal newPositionValue = currentPositionValue + change;

    return newPositionValue;
}

void SetRealizedPnL(StockState& stockState)
{
    if (stockState.position != 0)
    {
        throw runtime_error("Cannot set realized PnL because Position is not zero");
    }

    const auto percentage_denominator =
        GetDecimal(stockState.targetPosition + stockState.sharesPerInterval) *
        stockState.initialPrice;

    Decimal realizedPnLAsPercentage =
        (stockState.netPositionValue / percentage_denominator) * 100;

    stockState.realizedPnLAsPercentage = realizedPnLAsPercentage;
}

void UpdateSnaphotOnState(StockState& stockState, const Snapshot& snapshot)
{
    stockState.lastAsk = snapshot.ask;
    stockState.lastBid = snapshot.bid;
}

void UpdateExitPnL(StockState& stockState)
{
    const auto& lastAsk = stockState.lastAsk;
    const auto& lastBid = stockState.lastBid;
    const int position = stockState.position;

    if (position == 0)
    {
        return;
    }

    const string orderSide = position > 0 ? "SELL" : "BUY";

    const auto priceSetAt =
        (orderSide == static_cast<string>("BUY")) ? lastAsk : lastBid;

    const auto ifClosingPositionValue = GetNewNetPositionValue(
        stockState.netPositionValue,
        stockState.brokerageTradingCostPerShare,
        orderSide,
        0,
        position,
        priceSetAt
    );

    const auto percentage_denominator =
        GetDecimal(stockState.targetPosition + stockState.sharesPerInterval) *
        stockState.initialPrice;

    Decimal exitPnLAsPercentage =
        (ifClosingPositionValue / percentage_denominator) * 100;

    stockState.exitPnLAsPercentage = exitPnLAsPercentage;

    if (exitPnLAsPercentage > stockState.maxMovingProfitAsPercentage)
    {
        stockState.maxMovingProfitAsPercentage = exitPnLAsPercentage;
    }

    if (exitPnLAsPercentage < stockState.maxMovingLossAsPercentage)
    {
        stockState.maxMovingLossAsPercentage = exitPnLAsPercentage;
    }

    if (!stockState.reached_1_percentage_profit &&
        exitPnLAsPercentage >= GetDecimal(1.0))
    {
        stockState.reached_1_percentage_profit = true;
        stockState.max_loss_when_reached_1_percentage_profit =
            stockState.maxMovingLossAsPercentage;
    }

    if (!stockState.reached_0_75_percentage_profit &&
        exitPnLAsPercentage >= GetDecimal(0.75))
    {
        stockState.reached_0_75_percentage_profit = true;
        stockState.max_loss_when_reached_0_75_percentage_profit =
            stockState.maxMovingLossAsPercentage;
    }

    if (!stockState.reached_0_5_percentage_profit &&
        exitPnLAsPercentage >= GetDecimal(0.5))
    {
        stockState.reached_0_5_percentage_profit = true;
        stockState.max_loss_when_reached_0_5_percentage_profit =
            stockState.maxMovingLossAsPercentage;
    }

    if (!stockState.reached_0_25_percentage_profit &&
        exitPnLAsPercentage >= GetDecimal(0.25))
    {
        stockState.reached_0_25_percentage_profit = true;
        stockState.max_loss_when_reached_0_25_percentage_profit =
            stockState.maxMovingLossAsPercentage;
    }
}

void ReconcileRealizedPnLWhenHistoricalSnapshotsExhausted(StockState& stockState)
{
    UpdateExitPnL(stockState);

    SetNewPosition(
        stockState.brokerageId,
        stockState,
        0,
        Snapshot{stockState.lastAsk, stockState.lastBid}
    );

    SetRealizedPnL(stockState);
}

void CorrectBadBuyIfRequired(StockState& stockState, std::vector<int>& indexesToExecute)
{
    int lowestIndexExecuted = indexesToExecute.back();
    if (lowestIndexExecuted >= static_cast<int>(stockState.intervals.size()) - 1)
    {
        return;
    }

    auto& intervals = stockState.intervals;
    auto& intervalBelowLowestIntervalExecuted = intervals[lowestIndexExecuted + 1];
    if (!intervalBelowLowestIntervalExecuted.BUY.active)
    {
        return;
    }

    intervalBelowLowestIntervalExecuted.BUY.active = false;
    intervalBelowLowestIntervalExecuted.BUY.crossed = false;
    intervalBelowLowestIntervalExecuted.SELL.active = true;
    intervalBelowLowestIntervalExecuted.SELL.crossed = false;

    auto& topIntervalExecuted = intervals[indexesToExecute[0]];
    topIntervalExecuted.BUY.active = true;
    topIntervalExecuted.BUY.crossed = false;
    topIntervalExecuted.SELL.active = false;
    topIntervalExecuted.SELL.crossed = false;

    for (auto& interval : intervals)
    {
        interval.BUY.price = interval.BUY.price + stockState.spaceBetweenIntervals;
        interval.SELL.price = interval.SELL.price + stockState.spaceBetweenIntervals;
    }
}

void CorrectBadSellIfRequired(
    StockState& stockState, std::vector<int>& indexesToExecute
)
{
    int highestIndexExecuted = indexesToExecute[0];
    if (highestIndexExecuted == 0)
    {
        return;
    }

    auto& intervals = stockState.intervals;
    auto& intervalAboveHighestIntervalExecuted = intervals[highestIndexExecuted - 1];
    if (!intervalAboveHighestIntervalExecuted.SELL.active)
    {
        return;
    }

    intervalAboveHighestIntervalExecuted.SELL.active = false;
    intervalAboveHighestIntervalExecuted.SELL.crossed = false;
    intervalAboveHighestIntervalExecuted.BUY.active = true;
    intervalAboveHighestIntervalExecuted.BUY.crossed = false;

    auto& bottomIntervalExecuted = intervals[indexesToExecute.back()];
    bottomIntervalExecuted.SELL.active = true;
    bottomIntervalExecuted.SELL.crossed = false;
    bottomIntervalExecuted.BUY.active = false;
    bottomIntervalExecuted.BUY.crossed = false;

    for (auto& interval : intervals)
    {
        interval.BUY.price = interval.BUY.price - stockState.spaceBetweenIntervals;
        interval.SELL.price = interval.SELL.price - stockState.spaceBetweenIntervals;
    }
}

void AddSkippedBuysIfRequired(
    StockState& stockState, std::vector<int>& indexesToExecute
)
{
    if (indexesToExecute.empty())
    {
        return;
    }

    const auto& intervals = stockState.intervals;
    int bottomOriginalIndexToExecute = indexesToExecute.back();
    for (int i = intervals.size() - 1; i > bottomOriginalIndexToExecute; --i)
    {
        const auto& interval = intervals[i];

        if (interval.BUY.active)
        {
            indexesToExecute.push_back(i);
        }
    }
}

void AddSkippedSellsIfRequired(
    StockState& stockState, std::vector<int>& indexesToExecute
)
{
    if (indexesToExecute.empty())
    {
        return;
    }

    const auto& intervals = stockState.intervals;
    int topOriginalIndexToExecute = indexesToExecute[0];
    for (int i = 0; i < topOriginalIndexToExecute; ++i)
    {
        const auto& interval = intervals[i];

        if (interval.SELL.active)
        {
            indexesToExecute.insert(indexesToExecute.begin(), i);
        }
    }
}
